src/main/mondrian/xmla/XmlaHandler.java

```diff
@@ -2228,45 +2253,51 @@ private void axis(
         }
 
         private void writeMember(
-            SaxWriter writer,
-            Member member,
-            Position prevPosition,
-            Position nextPosition,
-            int k,
-            List<Property> props)
+            SaxWriter writer, Member member, Position prevPosition,
+            Position nextPosition, int k, List<Property> props)
             throws OlapException
         {
             writer.startElement(
-                "Member",
-                "Hierarchy", member.getHierarchy().getName());
-            for (Property prop : props) {
-                Object value;
-                Property longProp = longProps.get(prop.getName());
-                if (longProp == null) {
-                    longProp = prop;
-                }
+                "Member", "Hierarchy", member.getHierarchy().getName());
+            for (final Property prop : props) {
+                Object value = null;
+                Property longProp = (longProps.get(prop.getName()) != null)
+                    ? longProps.get(prop.getName()) : prop;
                 if (longProp == StandardMemberProperty.DISPLAY_INFO) {
-                    Integer childrenCard =
-                        (Integer) member.getPropertyValue(
+                    Integer childrenCard = (Integer) member
+                        .getPropertyValue(
                             StandardMemberProperty.CHILDREN_CARDINALITY);
                     value = calculateDisplayInfo(
-                        prevPosition,
-                        nextPosition,
-                        member, k, childrenCard);
+                        prevPosition, nextPosition, member, k, childrenCard);
                 } else if (longProp == StandardMemberProperty.DEPTH) {
                     value = member.getDepth();
                 } else {
-                    value = member.getPropertyValue(longProp);
+                    value = (longProp instanceof IMondrianOlap4jProperty)
+                        ? getCurrentHierarchyProperty(member, longProp)
+                        : member.getPropertyValue(longProp);
                 }
                 if (value != null) {
-                    writer.textElement(
-                        encoder.encode(prop.getName()), value);
+                    writer.textElement(encoder.encode(prop.getName()), value);
                 }
             }
-
             writer.endElement(); // Member
         }
 
+        private Object getCurrentHierarchyProperty(
+            Member member, Property longProp) throws OlapException
+        {
+            IMondrianOlap4jProperty currentProperty =
+                (IMondrianOlap4jProperty) longProp;
+            String thisHierarchyName = member.getHierarchy().getName();
+            String thatHierarchyName = currentProperty.getLevel()
+                .getHierarchy().getName();
+            if (thisHierarchyName.equals(thatHierarchyName)) {
```

Comment:
What if two hierarchies exist with the same name but within different dimensions? I think you should use the unique name of the hierarchy instead of just its name.